File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from itertools import combinations
from collections import Counter

def solution(orders, course):
    answer = []
    
    for num in course:
        combi = []
        for order in orders:
            order_list = sorted(list(order))
            combi += list("".join(p) for p in combinations(order_list, num))
        combi_dic = dict(Counter(combi))
        if(combi_dic):
            max_value = max(combi_dic.values())
            if max_value >= 2:
                for key, value in combi_dic.items():
                    if value == max_value:
                        answer.append(key)
        
    return sorted(answer)

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    public List<String> solution(String[] orders, int[] course) {
        List<String> answer = new ArrayList<>();
            
        for (int r : course) {
            Map<String, Integer> menus = new HashMap<>();

            for (String order : orders) {
                if (order.length() < r)
                    continue;
                
                char[] orderList = order.toCharArray();
                Arrays.sort(orderList);

                List<String> combos = combination(orderList, r);
                for (String combo : combos)
                    menus.put(combo, menus.getOrDefault(combo, 0)+1);
            }
            
            if (menus.isEmpty())
                continue;
            
            int maxValue = Collections.max(menus.values());
            if (maxValue >= 2) {
                for (Map.Entry<String, Integer> menu : menus.entrySet()) {
                    if (menu.getValue() == maxValue)
                        answer.add(menu.getKey());
                }
            }
        }
        
        Collections.sort(answer);
        return answer;
    }
    
    private List<String> combination(char[] list, int r) {
        List<String> result = new ArrayList<>();

        backtrack(list, result, new StringBuilder(), 0, r);
        
        return result;
    }
    
    private void backtrack(char[] list, List<String> result, StringBuilder sb, int start, int r) {
        if (sb.length() == r) {
            result.add(sb.toString());
            return;
        }
        
        for (int i = start; i < list.length; i++) {
            sb.append(list[i]);
            backtrack(list, result, sb, i + 1, r);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}